File: cs/competitive_programming/leetcode_fulltimes/2023/july/110_balanced_binary_tree.cpp
#include <iostream>

// TREE Example
/*
            x
        x       x
    x
x

*/

/**
 * Definition for a binary tree node.*/
struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode() : val(0), left(nullptr), right(nullptr) {}
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
    TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};


class Solution{

public:
    
    // tip: murali while practising wrote return type of below function as bool instead of int
    // so few tc failed, later he figured out the following issue. here his logic is correct but weird mistake.
    int helper(TreeNode* root) {

        if(root==nullptr) return 0;

        int l = this->helper(root->left);
        int r = this->helper(root->right);

        if( (l==-1) || (r==-1) || std::abs(l-r)>1) return -1;
        return 1+std::max(l,r);

    }

    bool isBalanced(TreeNode* root) {

        if(this->helper(root)==-1)
            return false;
        return true;
    }
};

File: cs/competitive_programming/leetcode_fulltimes/2023/januarary/110_balanced_binary_tree.py
# Definition for a binary tree node.
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right
class Solution:
    def isBalanced(self, root: Optional[TreeNode]) -> bool:


        def helper(node):

            if not node: return [True, 0]
            
            left = helper(node.left)
            right = helper(node.right)
            balanced = (left[0] and right[0] and abs(left[1]-right[1])<=1)
            h = 1+max(left[1],right[1])
            
            return [ balanced, h ]

        return helper(root)[0]

# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def isBalanced(self, root: Optional[TreeNode]) -> bool:

        def helper(node):

            if not node:
                return [True,0]
            
            l = helper(node.left)
            r = helper(node.right)

            balanced = (l[0] and r[0] and (abs(l[1]-r[1])<=1))
            h = 1+max(l[1],r[1])

            return [balanced, h]
        
        return helper(root)[0]

# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def isBalanced(self, root: Optional[TreeNode]) -> bool:

        res = True

        def helper(node):
            nonlocal res

            if not node:
                return 0
            
            l = helper(node.left)
            r = helper(node.right)

            res = (res and abs(l-r)<=1)
            return 1+max(l,r)
        
        helper(root)
        return res
